cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/KVTransformation.java

```diff
@@ -0,0 +1,179 @@
+import co.cask.cdap.etl.api.batch.BatchSink;
+import co.cask.cdap.etl.api.batch.BatchSource;
+import co.cask.cdap.etl.common.Constants;
+import co.cask.cdap.etl.common.DefaultEmitter;
+
+/**
+ * Key-Value transformation which wraps transformation for each batch plugin to emit stageName along with each record
+ * except for {@link BatchSink}
+ */
+public final class KVTransformation {
+
+  /**
+   * Creates {@link Transformation} which adds stageName to each record being emitted from current stage,
+   * except for {@link BatchSink}. Each stage will strip off stageName, apply original transformation and then again
+   * wrap record with stageName. For {@link co.cask.cdap.etl.api.Joiner} the record is passed with stageName.
+   *
+   * @param stageName      stageName which is emitting the record
+   * @param pluginType     type of the stage
+   * @param isFirstHalf    if first half of reduce type is being executed
+   * @param transformation transformation to be wrapped
+   * @return {@link Transformation} to wrap/unwrap stageName
+   */
+  public Transformation getKVTransformation(String stageName, String pluginType, boolean isFirstHalf,
+                                            Transformation transformation) {
+    if (BatchSink.PLUGIN_TYPE.equalsIgnoreCase(pluginType)) {
+      return new KVSinkTransformation<>(transformation);
+    } else if (BatchSource.PLUGIN_TYPE.equalsIgnoreCase(pluginType)) {
+      return new KVSourceTransformation<>(stageName, transformation);
+    } else if (Constants.CONNECTOR_TYPE.equalsIgnoreCase(pluginType)) {
+      // TODO Add SparkCompute and SparkSink once CDAP-6065 is merged
+      return transformation;
+    } else if (BatchJoiner.PLUGIN_TYPE.equalsIgnoreCase(pluginType)) {
+      if (isFirstHalf) {
+        return new KVJoinerTransformation<>(transformation);
+      } else {
+        return new KVSourceTransformation<>(stageName, transformation);
+      }
+    } else if (BatchAggregator.PLUGIN_TYPE.equalsIgnoreCase(pluginType)) {
+      if (isFirstHalf) {
+        return new KVSinkTransformation<>(transformation);
+      } else {
+        return new KVSourceTransformation<>(stageName, transformation);
+      }
+    }
+    return new KVWrappedTransformation(stageName, transformation);
+  }
+
+  /**
+   * Converts input to (stageName, intput)
+   *
+   * @param <IN>  type of input
+   * @param <OUT> type of output
+   */
+  public class KVSourceTransformation<IN, OUT> implements Transformation<IN, KeyValue<String, OUT>> {
+    private final String stageName;
+    private final Transformation<IN, OUT> transformation;
+
+
+    public KVSourceTransformation(String stageName, Transformation<IN, OUT> transformation) {
+      this.stageName = stageName;
+      this.transformation = transformation;
+    }
+
+    @Override
+    public void transform(IN input, Emitter<KeyValue<String, OUT>> emitter) throws Exception {
+      DefaultEmitter<OUT> singleEmitter = new DefaultEmitter<>();
```

Comment:
its a little better to create this emitter in the constructor, then reset it for each transform call.